src/zepben/evolve/streaming/grpc/grpc_channel_builder.py

```diff
@@ -23,16 +32,51 @@ def __init__(self):
         self._socket_address: str = "localhost:50051"
         self._channel_credentials: Optional[grpc.ChannelCredentials] = None
 
-    def build(self) -> grpc.aio.Channel:
+    def build(self, skip_connection_test: bool = False, debug: bool = False) -> grpc.aio.Channel:
         """
         Get the resulting :class:`grpc.aio.Channel` from this builder.
+
+        :param skip_connection_test: Skip confirming a connection can be established to the server. This is not recommended, but provided as a safety
+        mechanism if for any reason the connection test fails unexpectedly even though the connection is fine.
+        :param debug: Collect and append unhandled RPC errors to the `ConnectionException` raised on an unsuccessful connection test.
+
         :return: A gRPC channel resulting from this builder.
         """
+
+        # Synchronous channel is used purely for confirming channel connection works as we can't do async requests here,
+        # but an aio channel is always returned to the user.
+        if not skip_connection_test:
+            if self._channel_credentials:
+                channel = grpc.secure_channel(self._socket_address, self._channel_credentials)
+            else:
+                channel = grpc.insecure_channel(self._socket_address)
+            self._test_connection(channel, debug=debug)
+            channel.close()
+
         if self._channel_credentials:
             return grpc.aio.secure_channel(self._socket_address, self._channel_credentials)
 
         return grpc.aio.insecure_channel(self._socket_address)
 
+    def _test_connection(self, channel: grpc.Channel, debug: bool):
+        stubs: List = [
+            NetworkConsumerStub(channel),
+            DiagramConsumerStub(channel),
+            CustomerConsumerStub(channel)
+        ]
+        debug_errors = []
+        for client in stubs:
+            try:
+                result = client.getMetadata(GetMetadataRequest(), wait_for_ready=False)
+                if result:
+                    return
+            except _InactiveRpcError as rpc_error:
+                if rpc_error.code() in {grpc.StatusCode.UNKNOWN, grpc.StatusCode.UNAUTHENTICATED, grpc.StatusCode.UNAVAILABLE}:
+                    raise rpc_error
+                if debug:
+                    debug_errors.append(f"Received the following exception with {type(client).__name__}:\n{rpc_error}\n")
+        raise ConnectionException(f"Couldn't establish gRPC connection to any service on {self._socket_address}.\n{''.join(debug_errors) if debug else ''}")
+
```

Comment:
Should we also have the ability to check a subset of the services? e.g. we know that we will only use the network and diagram service, so we don't bother checking the customer service. This would be a fail-fast in the case where only the customer service is up, unlike this method that simply checks if *any* of the three services are up.

Reply:
The problem is we don't know what services the user will use. i.e, the connect functions get you a channel that you can then use with any client. The purpose of this is to just make sure that the channel can establish a connection, so we don't care which one it connects to, as long as it connects to at least one. This test is purely a connectivity test, rather than testing whether the specific client they want to use is supported by the server.